Enable custom project names

fvaresi opened the pull request:
This enables the user to customize the project name based on project-root. This is helpful in situations where the project base dir is not descriptive enough.

As an example, I use it like this:

```elisp
     (defun fvaresi/projectile-custom-project-name (project-root)
        (cond
         ((string-match "/Plugins/\\([^/]+\\)/\\(?:branches\\|tags\\)\\(?:.*\\)/\\([^/]+\\)" project-root)
          (let* ((product-name (match-string 1 project-root))
                 (branch-name (match-string 2 project-root)))
            (concat product-name "/" branch-name))) ;

         ((string-match "/Plugins/\\([^/]+\\)/trunk" project-root)
                (let* ((product-name (match-string 1 project-root)))
                  (concat product-name "/trunk")))
   
         ;; do the default
         (t
          (projectile-default-project-name project-root))))

      (setq projectile-project-name-function 'fvaresi/projectile-custom-project-name)
```

bbatsov commented on projectile.el:
```diff
@@ -207,6 +207,11 @@ Two example filter functions are shipped by default -
+(defcustom projectile-project-name-function 'projectile-default-project-name
+  "A function that receives the project-root and returns the project name to be displayed."
+  :group 'projectile
+  :type 'symbol)
```
This should have a package-version properly.

bbatsov commented on projectile.el:
```diff
@@ -736,6 +741,10 @@ A thin wrapper around `file-truename' that handles nil."
     (error nil)))
 
+(defun projectile-default-project-name (project-root)
+  "Default function used create project name to be displayed"
```
`M-x checkdoc` won't like this. `project-root` must appear in all-uppercase in the docstring and it should end with a `.`.

bbatsov commented:
This should be extended a bit further. If there's a buffer-local variable - `projectile-project-name` it should take precedence over calling this function IMO. The ultimate flexibility.

fvaresi commented:
Thanks for the feedback. I'll implement these changes as soon as I get a moment.

bbatsov commented:
You'll also have to rebase on top of the current `master`.

fvaresi commented:
I started implementing the precedence of a buffer-local-variable but saw some issues while testing this. I'm guessing this local variable would be for referencing projects in an external directory. Is this the use case you were thinking about?

I think using this would also require to use the variable proposed for customizing project-root (#329), otherwise it won't be able to find the project root.

Thanks again for your efforts and sorry about being so slow (these are my first baby steps in contributing to OS).

bbatsov commented:
> I think using this would also require to use the variable proposed for customizing project-root (#929), otherwise it won't be able to find the project root.

It's totally unrelated. The idea is to be able to set the project name via `.dir-locals.el`, without having to deal with custom name functions. Extremely useful if you don't really want to derive the project name from some directory, but would rather set it explicitly.

fvaresi commented:
I now see what you meant and implemented the suggested changes. I think tests are currently giving an error cause helm-projectile was moved to its own repo.

Also, since persp-projectile was also moved I'll send a new pull request for changes related to this one, once this pull request is approved.

bbatsov commented:
The changes look good. Just squash the 3 commits together and we're good to go.

fvaresi commented:
It's done, thanks.

bbatsov commented:
And drop that `.` from the commit message - titles shouldn't end with `.`.

fvaresi commented:
Ok, modified commit message.

bbatsov commented:
:+1: Thanks!

ReneFroger commented:
@fvaresi could you tell more how you could specifice a custom project name inside a project? I couldn't figure it out from your code and the changes.

fvaresi commented:
@ReneFroger if the custom project name is a function depending on the project root, you can set `projectile-project-name-function` as shown in the first comment of this pull request.

If you want to customize the project name just for one project, it's easier to set the local variable `projectile-project-name` using [directory variables](https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Variables.html). This is probably the simplest way, since you just need to add a `.dir-locals.el` file in the project root.

ReneFroger commented:
Thanks for your answer, but I got the asking message everytime:

    "The local variables in [project folder] contains values that may not be safe (*)

    Do you want to apply it? 

    y
    n
    !

    * projectile-project-name . "my-super-project"

Is there any way to turn that confirmation message off? Typing ! didn't help
after I visit the project again.

fvaresi commented:
In my setup using ! will just add one safe value. If you want to say that all strings are safe values to be used you can add this to your emacs config:

```emacs-lisp
(put 'projectile-project-name 'safe-local-variable #'stringp)
```

ReneFroger commented:
@fvaresi thanks for your helpful response. It works like a charm, and integrates well with my yasnippets (https://github.com/bbatsov/projectile/issues/848), thanks for your contribution, it's appreciated!
